### ISSUE TITLE
Memory leak in Vulnerability Detector

### ISSUE THREAD

**Issue by vikman90** [wazuh/wazuh#2616]
|Branch|Install type|Install method|Platform|
|---|---|---|---|
| v3.8.1 / 3.9 | Manager | Sources | CentOS 7 |

## Valgrind report

```
==27520== 222,912 (120,008 direct, 102,904 indirect) bytes in 1 blocks are definitely lost in loss record 199 of 202
==27520==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==27520==    by 0x54F1976: sqlite3MemMalloc (sqlite3.c:20790)
==27520==    by 0x54F22F4: mallocWithAlarm (sqlite3.c:24464)
==27520==    by 0x54F238B: sqlite3Malloc (sqlite3.c:24494)
==27520==    by 0x5589CF2: setupLookaside (sqlite3.c:141147)
==27520==    by 0x558CF53: openDatabase (sqlite3.c:143531)
==27520==    by 0x558D035: sqlite3_open_v2 (sqlite3.c:143595)
==27520==    by 0x44DDDA: wm_vuldet_create_file (wm_vuln_detector.c:177)
==27520==    by 0x4510A1: wm_vuldet_check_db (wm_vuln_detector.c:1083)
==27520==    by 0x44FB1F: wm_vuldet_check_agent_vulnerabilities (wm_vuln_detector.c:746)
==27520==    by 0x457AEF: wm_vuldet_main (wm_vuln_detector.c:2480)
==27520==    by 0x5A38DD4: start_thread (in /usr/lib64/libpthread-2.17.so)
==27520==    by 0x5D4AEAC: clone (in /usr/lib64/libc-2.17.so)
==27520==
==27520== 222,912 (87,208 direct, 135,704 indirect) bytes in 1 blocks are definitely lost in loss record 200 of 202
==27520==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==27520==    by 0x54F1976: sqlite3MemMalloc (sqlite3.c:20790)
==27520==    by 0x54F22F4: mallocWithAlarm (sqlite3.c:24464)
==27520==    by 0x54F238B: sqlite3Malloc (sqlite3.c:24494)
==27520==    by 0x54FF58B: pcache1InitBulk (sqlite3.c:45289)
==27520==    by 0x54FF8A9: pcache1AllocPage (sqlite3.c:45409)
==27520==    by 0x550032C: pcache1FetchStage2 (sqlite3.c:45902)
==27520==    by 0x55004A6: pcache1FetchNoMutex (sqlite3.c:46006)
==27520==    by 0x55004D5: pcache1Fetch (sqlite3.c:46048)
==27520==    by 0x54FE921: sqlite3PcacheFetch (sqlite3.c:44511)
==27520==    by 0x5505B1B: getPageNormal (sqlite3.c:52332)
==27520==    by 0x5505FC4: sqlite3PagerGet (sqlite3.c:52516)
==27520==    by 0x550E67D: btreeGetPage (sqlite3.c:61025)
==27520==    by 0x550FA82: lockBtree (sqlite3.c:61945)
==27520==    by 0x551020D: sqlite3BtreeBeginTrans (sqlite3.c:62308)
==27520==    by 0x556036A: sqlite3InitOne (sqlite3.c:115848)
==27520==    by 0x5560800: sqlite3Init (sqlite3.c:116027)
==27520==    by 0x55608EF: sqlite3ReadSchema (sqlite3.c:116064)
==27520==    by 0x5547466: sqlite3StartTable (sqlite3.c:100602)
==27520==    by 0x5583A31: yy_reduce (sqlite3.c:137725)
==27520==
==27520== 227,016 (680 direct, 226,336 indirect) bytes in 1 blocks are definitely lost in loss record 201 of 202
==27520==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==27520==    by 0x54F1976: sqlite3MemMalloc (sqlite3.c:20790)
==27520==    by 0x54F22F4: mallocWithAlarm (sqlite3.c:24464)
==27520==    by 0x54F238B: sqlite3Malloc (sqlite3.c:24494)
==27520==    by 0x54F2AE6: sqlite3MallocZero (sqlite3.c:24799)
==27520==    by 0x558CAB1: openDatabase (sqlite3.c:143328)
==27520==    by 0x558D035: sqlite3_open_v2 (sqlite3.c:143595)
==27520==    by 0x44DDDA: wm_vuldet_create_file (wm_vuln_detector.c:177)
==27520==    by 0x4510A1: wm_vuldet_check_db (wm_vuln_detector.c:1083)
==27520==    by 0x44FB1F: wm_vuldet_check_agent_vulnerabilities (wm_vuln_detector.c:746)
==27520==    by 0x457AEF: wm_vuldet_main (wm_vuln_detector.c:2480)
==27520==    by 0x5A38DD4: start_thread (in /usr/lib64/libpthread-2.17.so)
==27520==    by 0x5D4AEAC: clone (in /usr/lib64/libc-2.17.so)
==27520==
```

**Comment by vikman90** [comment]
Valgrind only reports these memory leaks if we close Modulesd during a Syscollector scan. Instead, if we let Syscollector scan finish, no leaks are reported.

I wonder if the low-level Berkeley DB data structure may be making some pointer references hard to track.

**Comment by crolopez** [comment]
There are 2 more memory leaks:

```
==17402== 53,560 bytes in 2,678 blocks are definitely lost in loss record 119 of 129
==17402==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17402==    by 0x17B300: wm_vuldet_json_parser (wm_vuln_detector.c:2147)
==17402==    by 0x179402: wm_vuldet_update_feed (wm_vuln_detector.c:1682)
==17402==    by 0x172524: wm_vuldet_check_update (wm_vuln_detector.c:165)
==17402==    by 0x17A798: wm_vuldet_run_update (wm_vuln_detector.c:2009)
==17402==    by 0x17AA90: wm_vuldet_updatedb (wm_vuln_detector.c:2044)
==17402==    by 0x17CDBA: wm_vuldet_main (wm_vuln_detector.c:2469)
==17402==    by 0x58686DA: start_thread (pthread_create.c:463)
==17402==    by 0x5BA188E: clone (clone.S:95)
==17402== 
==17402== 115,632 bytes in 27,346 blocks are definitely lost in loss record 121 of 129
==17402==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17402==    by 0x1967AA: OSRegex_Execute_ex (os_regex_execute.c:119)
==17402==    by 0x195EC7: OSRegex_Execute (os_regex_execute.c:30)
==17402==    by 0x1798F0: wm_vuldet_decode_package_version (wm_vuln_detector.c:1767)
==17402==    by 0x17B5C2: wm_vuldet_json_parser (wm_vuln_detector.c:2157)
==17402==    by 0x179402: wm_vuldet_update_feed (wm_vuln_detector.c:1682)
==17402==    by 0x172524: wm_vuldet_check_update (wm_vuln_detector.c:165)
==17402==    by 0x17A798: wm_vuldet_run_update (wm_vuln_detector.c:2009)
==17402==    by 0x17AA90: wm_vuldet_updatedb (wm_vuln_detector.c:2044)
==17402==    by 0x17CDBA: wm_vuldet_main (wm_vuln_detector.c:2469)
==17402==    by 0x58686DA: start_thread (pthread_create.c:463)
==17402==    by 0x5BA188E: clone (clone.S:95)
```

Fixed in 6d59d5ff0de0b0b6ec46ab2f510474ebd6976379.
